### PR TITLE
Fetch remote status on behalf of the user pasting URL (fixes #8193)

### LINKED ISSUE #8193: Fetch toots on behalf of the user requesting them in the search box
Remote toots (and accounts) can be manually fetched by pasting their URL in the search box.

However, it fetches them anonymously, thus, pasting the URL of a remote private or direct toot will always fail.

This should be easily fixable by fetching the toot on behalf of the user pasting the URL. There are some minor privacy concerns with that, though: until now, the remote server did not know which user requested to fetch a given toot. This could be an optional feature (either opt-in or opt-out, I don't know), or this concern might be slightly reduced by attempting to get the toot anonymously first, which will cause additional HTTP requests in case of private/direct toots.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Since it can disclose to the remote instance that the local user is trying to fetch a toot, that is opt-in:
screenshot_2018-08-19 preferences - dev instance 3

EDIT: Edited to reflect that it is now opt-in.

If the feature hasn't been opted in and an URL-based search fails to return a result, the following warning is displayed, linking to the above setting:
screenshot_2018-08-19 dev instance 5

### CODE REVIEW

**Comment by ThibG** [comment]
I am wondering if it wouldn't be better to try once anonymously, then retry on behalf of the account if the result is not 200 OK.
This would “leak” the information in less circumstances and be slightly less expensive in the most common case, at the cost of one additional HTTP request when the remote resource *is* private.

**Review by Gargron — CHANGES_REQUESTED** [review]

File: config/locales/simple_form.en.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ en:
         phrase: Will be matched regardless of casing in text or content warning of a toot
         scopes: Which APIs the application will be allowed to access. If you select a top-level scope, you don't need to select individual ones.
         setting_default_language: The language of your toots can be detected automatically, but it's not always accurate
+        setting_fetch_anonymously: Unless this is checked, fetching a remote toot or user from the search box will let the remote server know about your account. If checked, you won't be able to fetch private toots.
```
[the remark this answers is not in the document]
Comment (ThibG) [reply config/locales/simple_form.en.yml]: I'll rewrite it, it's slightly misleading now that the next commit only authenticates you to the remote server when fetching first failed.

File: config/settings.yml [hunk 2]
```diff
@@ -22,6 +22,7 @@ defaults: &defaults
   show_staff_badge: true
   default_sensitive: false
   hide_network: false
+  fetch_anonymously: false
```
[the remark this answers is not in the document]
Comment (ThibG) [reply config/settings.yml]: I'm torn about this one. It's unlikely to happen and is only an issue if pasting an URL of a remote resource stored on a malicious server.
On the contrary, I think this is useful to far more people than it is an issue for.
However, it *is* a change of behavior and privacy expectations, however small it is, so…

**Comment by ThibG** [comment]
There is an alternative, more involved way to go about how to consent to disclosing the account name to remote instances:
1. Have the setting be on by default, but renamed to something like “always ask”
2. If anonymously getting the remote status failed, and the setting is on, display a message like “The resource you requested might require authentication. Retry by disclosing your account name to the remote instance?”

**Comment by Gargron** [comment]
@ThibG That sounds way too complicated to be worth it.

**Comment by ThibG** [comment]
@Gargron to me, that sounds like the best way to do it. But we can implement that later if the feature is opt-out. I will make the setting on by default.

**Comment by ThibG** [comment]
I have made the changes you requested.

**Review by nightpool — CHANGES_REQUESTED** [review]
making this an option is a huge anti-pattern. If we think this is a security/privacy risk, we should make sure to foreground it to a user *when they fetch the status*, not in a hidden preferences checkbox that nobody will ever see.

**Comment by nightpool** [comment]
Alternatively, i think taking out the preferences checkbox and doing the optimistic-anonymous and retrying with authorization covers 90% of the possible security problems.

**Comment by ThibG** [comment]
Not completely convinced by the color scheme, but here is what I have now:
screenshot_2018-08-19 dev instance 3

**Comment by ThibG** [comment]
So… addressed @Gargron's concerns by setting `fetch_anonymously` to `true` by default and changing the wording, and addressed @nightpool's concerns by making the setting discoverable via a warning appearing if the setting is likely to have prevented fetching a private toot.

**Comment by trwnh** [comment]
"authentify"? i think you mean "authenticate" or "identify"

**Comment by ThibG** [comment]
@trwnh fixed!

**Comment by trwnh** [comment]
Revisiting this after I just now had some difficulty sharing an old followers-only status with one of my followers... but I'd like to clarify the flow of what exactly is going on here.

I'm not sure the current setting makes sense as I understand it. Is it opt-out or opt-in? @ThibG's comment says `fetch_anonymously` is set to true by default, and the screenshot shows a "change this setting" link, but the message for commit https://github.com/tootsuite/mastodon/pull/8214/commits/33b401a48417bb009ac3047505691c1a685e3e74 says "allow users opting out". So, is it currently opt-in or opt-out for this PR?

**Comment by ThibG** [comment]
Sorry this is a bit confusing. This depends on whether you consider the feature to be “stay anonymous” or “fetch private toots”.
The setting is called “always fetch anonymously” and is on by default. This means you won't be able to fetch remote private toots by default.
You can then, by disabling this feature, opt-in to see private toots at the cost of occasionally disclosing to the remote server that *you* are accessing a given resource from them.

Le 20 septembre 2018 04:41:37 GMT+02:00, trwnh <notifications@github.com> a écrit :
>Revisiting this after I just now had some difficulty sharing an old
>followers-only status with one of my followers... but I'd like to
>clarify the flow of what exactly is going on here.
>
>I'm not sure the current setting makes sense as I understand it. Is it
>opt-out or opt-in? @ThibG's comment says `fetch_anonymously` is set to
>true by default, and the screenshot shows a "change this setting" link,
>but the message for commit
>https://github.com/tootsuite/mastodon/pull/8214/commits/33b401a48417bb009ac3047505691c1a685e3e74
>says "allow users opting out". So, is it currently opt-in or opt-out
>for this PR?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/tootsuite/mastodon/pull/8214#issuecomment-423021361

-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by kit-ty-kate** [comment]
What is the status of this PR? It's really annoying to not being able to give a link to a private toot to someone.

**Comment by kit-ty-kate** [comment]
Can this be reopened?

**Comment by ThibG** [comment]
I am not actually completely sure we want something like this, and Pleroma, for instance, isn't going to offer fetching private toots as far as I understand.

However, being able to refer to private toots by URL—at least for those already known to the instance—is still valuable. To this end, I guess we could fetch toots from the database by their `url` attribute. One issue with that is that we can't ensure this URL actually corresponds to the private toot, and have to trust the originating instance. But that assumption is consistent with the other ones that we make regarding the instance being authoritative for every object living on their domain name.

**Comment by tsmethurst** [comment]
Hiya, are there any plans to reconsider / update this? Just linking to this issue here since it's related: https://github.com/superseriousbusiness/gotosocial/issues/3177

**Pull request closed without merging**